Added split documentation from main repo

GeeH opened the pull request:
This is just a first stab, but I have a script that can do this now - @weierophinney does this look right to you? I haven't tried rendering with Bookdown yet.

GeeH commented:
This renders now using Bookdown - it looks right to me @weierophinney.

danizord commented on doc/bookdown.json:
```diff
@@ -0,0 +1 @@
+{"title":"Zend\\Authentication","target":"html\/","content":["book\/zend.authentication.intro.md","book\/zend.authentication.adapter.dbtable.md","book\/zend.authentication.adapter.digest.md","book\/zend.authentication.adapter.http.md","book\/zend.authentication.adapter.ldap.md","book\/zend.authentication.validator.authentication.md"]}
```
Can u make this file more readable? Line breaks, indentation...

weierophinney replied:
I agree; we'll need to be able to edit it in order to insert or re-order pages.

GeeH replied:
:+1:

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+splitted off into `Zend\Authentication\Adapter\DbTable\CallbackCheck` and
+`Zend\Authentication\Adapter\DbTable\CredentialTreatmentAdapter`. Use
+`Zend\Authentication\Adapter\DbTable\CredentialTreatmentAdapter` instead of
+`Zend\Authentication\Adapter\DbTable`.
```
Each of the above lines needs to be prefixed with `>` to be part of the blockquote for the note.

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+# Database Table Authentication
+
+> **note**
```
Let's rewrite this as "Deprecation Notice" or similar.

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+
+The available configuration options include:
+
+-   **tableName**: This is the name of the database table that contains the authentication
```
I'm guessing the extra space between the `-` and the text was due to pandoc. We may need to write a post-processor to fix stuff like that.

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+$dbAdapter = new DbAdapter(array(
+                'driver' => 'Pdo_Sqlite',
+                'database' => 'path/to/sqlite.db'
+            ));
```
Since we have the opportunity, we should likely reformat these to follow PSR-1/2.

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+By default, `Zend\Authentication\Adapter\DbTable` returns the identity supplied back to the auth
+object upon successful authentication. Another use case scenario, where developers want to store to
+the persistent storage mechanism of `Zend\Authentication` an identity object containing other useful
+information, is solved by using the `getResultRowObject()` method to return a **stdClass** object.
```
`stdClass` should be in code blocks, not emphasis.

weierophinney replied:
That was likely wrong in the sources, though. Really sticks out, now!

weierophinney commented on doc/book/zend.authentication.adapter.dbtable.md:
```diff
@@ -0,0 +1,304 @@
+>
+> You can improve security even more by using a static salt value hard coded into your application.
+In the case that your database is compromised (e. g. by an *SQL* injection attack) but your web
+server is intact your data is still unusable for the attacker.
```
Same notes as for the initial note; let's get an improved title, and make sure all lines are prefixed with `>`.

weierophinney commented on doc/book/zend.authentication.adapter.digest.md:
````diff
@@ -0,0 +1,64 @@
+The above elements are separated by colons, as in the following example (in which the password is
+"**somePassword**"):
+
+``` {.sourceCode .text}
````
Flow this onto the next line.

weierophinney commented on doc/book/zend.authentication.adapter.http.md:
```diff
@@ -0,0 +1,149 @@
+is supposed to determine the valid lifetime of a given nonce, or effectively how long a client's
+authentication information is accepted. Currently, if it's set to 3600 (for example), it will cause
+the adapter to prompt the client for new credentials every hour, on the hour. This will be resolved
+in a future release, once nonce tracking and stale support are implemented.
```
Same comments as for other notes.

weierophinney commented on doc/book/zend.authentication.adapter.http.md:
````diff
@@ -0,0 +1,149 @@
+also be passed to the constructor. Its `resolve()` method walks through the text file, searching for
+a line with a matching username and realm. The text file format similar to Apache htpasswd files:
+
+``` {.sourceCode .text}
````
Flow onto next line.

weierophinney commented on doc/book/zend.authentication.adapter.ldap.md:
```diff
@@ -0,0 +1,248 @@
+The `Zend\Authentication\Adapter\Ldap` constructor accepts three parameters.
+
+The `$options` parameter is required and must be an array containing one or more sets of options.
+Note that it is **an array of arrays** of Zend\\\\Ldap\\\\Ldap \<zend.ldap.introduction\> options.
```
Whoa… escaping gone wild! I think this is a problem due to inter-document linking. We'll have to do a little work around those.

weierophinney commented on doc/book/zend.authentication.adapter.ldap.md:
```diff
@@ -0,0 +1,248 @@
+components are evaluated in reverse. For example, the user account **CN=Bob
+Carter,CN=Users,DC=w,DC=net** is located directly within the **CN=Users,DC=w,DC=net container**.
+This structure is best explored with an *LDAP* browser like the *ADSI* Edit *MMC* snap-in for Active
+Directory or phpLDAPadmin.
```
Same comment as for previous notes.

weierophinney commented on doc/book/zend.authentication.adapter.ldap.md:
```diff
@@ -0,0 +1,248 @@
+next set of server options is tried. If the bind is successful, the iteration stops, and the
+adapter's `authenticate()` method returns a successful result. If all server options have been tried
+without success, the authentication fails, and `authenticate()` returns a failure result with error
+messages from the last iteration.
```
Same comment as for previous notes.

weierophinney commented on doc/book/zend.authentication.intro.md:
```diff
@@ -0,0 +1,408 @@
+-   `isValid()`- returns `TRUE` if and only if the result represents a successful authentication
+attempt
+-   `getCode()`- returns a `Zend\Authentication\Result` constant identifier for determining the type
+of authentication failure or whether success has occurred. This may be used in situations where the
```
Lines that belong to a bullet point need to be indented two spaces. (Again, may need to do a post-processor.)

weierophinney commented on doc/book/zend.authentication.intro.md:
```diff
@@ -0,0 +1,408 @@
+`Zend\Authentication\AuthenticationService::authenticate()` stores the identity from the
+authentication result into persistent storage. Unless specified otherwise,
+`Zend\Authentication\AuthenticationService` uses a storage class named
+`Zend\Authentication\Storage\Session`, which, in turn, uses Zend\\\\Session \<zend.session\>. A
```
Another inter-document link.

weierophinney commented on doc/book/zend.authentication.intro.md:
```diff
@@ -0,0 +1,408 @@
+`Zend\Authentication\Storage\Session` uses a session namespace of '`Zend_Auth`'. This namespace may
+be overridden by passing a different value to the constructor of
+`Zend\Authentication\Storage\Session`, and this value is internally passed along to the constructor
+of Zend\\\\Session\\\\Container \<zend.session\>. This should occur before authentication is
```
Inter-document link.

weierophinney commented:
@GeeH This is _awesome_!

Here's a list of general issues I see in the conversion:
1. Notes are not formatted properly. The first line of the first paragraph receives the correct `>` prefix, but subsequent lines do not.
   
   Additionally, we need to come up with descriptive titles for the notes, and use actual headings (i.e., `###`) for each.
2. Bullet points. The correct prefix is used, but too many spaces are used between the delimiter and the first line of text. Additionally, lines are not flowed properly; they should be indented an additional 2 spaces, but are not.
3. Many acronyms are using `*ACRONYM*` style notation; those can likely go when we edit.
4. Inter-document linking does not translate betwee rst and markdown.
5. A _lot_ of examples are not following PSR-1/2. We should try and fix those if it's not too much effort.
6. Bookdown.json should be human-readable, as it will be edited by humans in order to insert new pages or re-order pages.

My guess is that most of the formatting issues are due to the conversion from rst to markdown via pandoc. If so, we may be able to solve points 1, 2, and 5 by post-processing the results:
1. We'd look for a line beginning with `>`, and check each following line; if the line is non-empty, we prefix it. For the first line, if the text matches `/\*\*([^*]+)\*\*/`, we change it to `### $1`.
2. We'd do a regex for `^(\s*)-\s+` and replace it with `$1-`. Similar to (1), we'd also check each following line, and if it's non-empty and does not start with whitespace or a `-`, we indent with `$1`.
3. We can take the contents of a PHP fenced code block and run it through `php-cs-fixer` or `phpcbf`, and then re-inject the results into the block.

Point 3 we _might_ be able to handle via regex: `s/\*([A-Z-_]+)\*/$1/`, but we'll need to spot-check.

Point 4 will be harder. I'm not entirely sure how we'll handle it, to be honest, until we start building the master manual. We could potentially link these to where the **markdown** files will live in each repo, however: eg., `[Zend\Ldap](https://github.com/zendframework/zend-ldap/blob/master/doc/book/zend.ldap.md)`. If we go this route, we might be able to _pre-process_ the file to change those links.

As for 6, pass it through `jq` and dump the results. :)

GeeH commented:
Agreed, I'll look at modifying the script for 1,2 and 5 tomorrow and resubmit.

GeeH commented:
Script now adds human readable JSON - thanks @danizord

GeeH commented:
1. Fixed
2. Fixed
3. I think we need to raise issues for each repo to say "fix acronyms in this format and make them sane" then hope people create the PR to fix them
4. I don't actually know what we need to do here. Again, I think we should raise standard issues in the docs to fix this by eye but not until we have relative urls for when the docs are rendered. I propose we ignore for now
5. Working on it
6. Fixed

GeeH commented:
I've done 5 to the best of my ability.

@weierophinney - what do you think about 3 and 4 please?

weierophinney commented:
@GeeH Looks great!

I agree with your assessment of my third point: 

> I think we need to raise issues for each repo to say "fix acronyms in this format and make them sane" then hope people create the PR to fix them

I also think this applies to 5 (as you and I determined that most of the formatting problems are not handled by either php-cs-fixer or phpcbf at this time).

As for 4, I agree. However, I'd like to do the following:
- Pre-process such links to become normal markdown link syntax.
- Raise an issue in the repo to fix those links once the docs for the relevant components are created.

I think this will ensure we can get them fixed well in the future. Since they follow a standard syntax, you should be able to pre-process them easily:

``` php
$contents = preg_replace('/:ref:`(?P<text>[^<]+)<(?P<link>[^>]+)>`/s', '[$1]($2)', $contents);
```

GeeH commented:
Roger that... on it.

GeeH commented:
Damn you to Hades with your **pre** processing

GeeH commented:
Ok, that's now done @weierophinney  - what's next?

weierophinney commented on doc/book/zend.authentication.intro.md:
```diff
@@ -0,0 +1,427 @@
+`Zend\Authentication\AuthenticationService::authenticate()` stores the identity from the
+authentication result into persistent storage. Unless specified otherwise,
+`Zend\Authentication\AuthenticationService` uses a storage class named
+`Zend\Authentication\Storage\Session`, which, in turn, uses [Zend\\Session ](zend.session). A custom
```
Crap. Looks like we need to adjust the PCRE to:

``` php
$docs = preg_replace('/:ref:`(.*?) <([^>]+)>`/', '[$1]($2)', $docs);
```

(links have the sequence `<`; previous regex wasn't taking into account the whitespace before the `<`, which leads to extra whitespace inside the link text).

weierophinney commented:
> Ok, that's now done @weierophinney - what's next?

Once the regex is corrected for the links and you're re-run and pushed, I'll merge this. Then it's a matter of moving on to the next repos. :)

GeeH commented:
Looking better :sunglasses:
